[StructuralMechanicsApplication] Harmonic Analysis Strategy - Moving BuildRHS() from Initialize() to SolveSolutionStep() in order to be able to do CoSimulation

juancamarotti opened the pull request:
## Summary  
This PR moves the `BuildLHS()` call from `Initialize()` to `SolveSolutionStep()` to enable **CoSimulation**.  You can see the change in [`harmonic_analysis_strategy.hpp`](https://github.com/KratosMultiphysics/Kratos/commit/38de9c68bbba65313677812d5b72f65dd459da94#diff-a60f91cac211bbc2c44fd30ca99b28b7aeac44c77a09275d1c8c9d451a77ddb9).

## Reason 
Previously, the **RHS** was not updated correctly after data transfer from the other solver.  
By relocating `BuildLHS()` to `SolveSolutionStep()`, we ensure that the **RHS is updated accordingly**, improving consistency in the simulation process.  

FYI: @matekelemen @sunethwarna @rickyaristio

juancamarotti committed:
Bugfix in harmonic_analysis_strategy.hpp

We move the BuildLHS() call from Initialize() to SolveSolutionStep() in order to be able to do CoSimulation

matekelemen commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/harmonic_analysis_strategy.hpp:
```diff
@@ -245,7 +245,6 @@ class HarmonicAnalysisStrategy
             if (r_force_vector.size() != system_size)
                 r_force_vector.resize(system_size, false);
             r_force_vector = ZeroVector( system_size );
```
```suggestion
            TSparseSpace::SetToZero(r_force_vector);
```

juancamarotti committed:
Changing r_force_vector initialization

Co-authored-by: Máté Kelemen <44344022+matekelemen@users.noreply.github.com>

sunethwarna reviewed (dismissed):
Nice catch @juancamarotti. I am ok as long as the unit tests passes. :)

juancamarotti committed:
Adding BuildRHS() in the Initialize()

juancamarotti committed:
Update harmonic_analysis_strategy.hpp

juancamarotti committed:
Update harmonic_analysis_strategy.hpp

juancamarotti committed:
Reverting the changes

juancamarotti committed:
Solving a bug in SolveSolutionStep()

juancamarotti committed:
Changing not with !

juancamarotti committed:
Build RHS() manually

In this new commit, the RHS() is computed manually by looping over the conditions within SolveSolutionStep(), in order to bypass the residual-based formulation used by the BuilderAndSolver (which is giving wrong results)

juancamarotti committed:
Adding EquationIdVectorType

juancamarotti committed:
Update harmonic_analysis_strategy.hpp

sunethwarna commented on applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/harmonic_analysis_strategy.hpp:
```diff
@@ -393,6 +394,21 @@ class HarmonicAnalysisStrategy
+            for (IndexType i = 0; i < equation_id.size(); ++i) {
+                     f[equation_id[i]] += rhs[i];
+            }
+        }
```
```suggestion
        IndexPartition<std::size_t>(r_model_part.NumberOfConditions()).for_each(std::tuple<DenseVectorType, EquationIdVectorType>(), [&r_model_part](const auto Index, auto& rTLS) {   
            auto& r_condition = *(r_model_part.ConditionsBegin() + Index);
            auto& rhs = std::get<0>(rTLS);
            auto& equation_id = std::get<1>(rTLS);
            
            r_condition.CalculateRightHandSide(rhs, r_process_info);
            r_condition.EquationIdVector(equation_id, r_process_info);

            // Assemble into global vector
            for (IndexType i = 0; i < equation_id.size(); ++i) {
                     AtomicAdd<double>(f[equation_id[i]], rhs[i]);
            }
        });
```

sunethwarna reviewed (commented):
I would do the load vector assembly in parallel

juancamarotti committed:
Parallelizing the RHS (force vector) assembly

sunethwarna reviewed (approved):
Thanks @juancamarotti :)
